core/presence/presence.go

```diff
@@ -0,0 +1,343 @@
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.enabled
+}
+
+func (r *recorder) find(server string, id uint64) int {
+	for i, e := range r.entries {
+		if e.Server == server && e.ConnectionID == id {
+			return i
+		}
+	}
+	return -1
+}
+
+// Connect adds an entry for the specified agent.
+func (r *recorder) Connect(server, model, agent string, id uint64, controllerAgent bool, userData string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	pos := r.find(server, id)
+	if pos == -1 {
+		r.entries = append(r.entries, Value{
+			Model:           model,
+			Server:          server,
+			Agent:           agent,
+			ControllerAgent: controllerAgent,
+			ConnectionID:    id,
+			Status:          Alive,
+			UserData:        userData,
+			LastSeen:        r.clock.Now(),
+		})
+	} else {
+		// Need to access the value in the array, not a copy of it.
+		r.entries[pos].Status = Alive
+		r.entries[pos].LastSeen = r.clock.Now()
+	}
+}
+
+// Disconnect removes the entry for the specified connection id.
+func (r *recorder) Disconnect(server string, id uint64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	if pos := r.find(server, id); pos >= 0 {
+		if pos == 0 {
+			r.entries = r.entries[1:]
+		} else {
+			r.entries = append(r.entries[0:pos], r.entries[pos+1:]...)
+		}
+	}
+}
+
+// Activity updates the last seen timestamp for the connection specified.
+func (r *recorder) Activity(server string, id uint64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	if pos := r.find(server, id); pos >= 0 {
+		r.entries[pos].LastSeen = r.clock.Now()
+	}
+}
+
+// ServerDown marks all connections on the specified server as unknown.
+func (r *recorder) ServerDown(server string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	for i, value := range r.entries {
+		if value.Server == server {
+			r.entries[i].Status = Missing
+		}
+	}
+}
+
+// UpdateServer replaces all known connections for the specified server with
+// the connections specified. If any of the  connections are not for the
+// specified server, and error is returned.
+func (r *recorder) UpdateServer(server string, connections []Value) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return errors.New("recorder not enabled")
+	}
+
+	entries := make([]Value, 0, len(r.entries))
+	for _, value := range r.entries {
+		if value.Server != server {
+			entries = append(entries, value)
+		}
+	}
+
+	for _, value := range connections {
+		if value.Server != server {
+			return errors.Errorf("connection server mismatch, got %q expected %q", value.Server, server)
+		}
+		value.Status = Alive
+		value.LastSeen = r.clock.Now()
+		entries = append(entries, value)
+	}
+
+	r.entries = entries
+	return nil
+}
+
+// Connections returns all connections info that the recorder has.
+func (r *recorder) Connections() Connections {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	entries := make([]Value, len(r.entries))
```

Comment:
no enabled check?

Reply:
If it isn't enabled, it'll return an empty slice.